src/switch_rtp.c

```diff
@@ -5882,6 +5890,92 @@ static switch_size_t do_flush(switch_rtp_t *rtp_session, int force, switch_size_
 				bytes = sizeof(rtp_msg_t);
 				switch_socket_recvfrom(rtp_session->from_addr, rtp_session->sock_input, 0, (void *) &rtp_session->recv_msg, &bytes);
 
+#ifdef ENABLE_SRTP
+
+            if (rtp_session->flags[SWITCH_RTP_FLAG_SECURE_RECV] && rtp_session->has_rtp &&
+                (check_recv_payload(rtp_session) ||
+                 rtp_session->last_rtp_hdr.pt == rtp_session->recv_te ||
+                 rtp_session->last_rtp_hdr.pt == rtp_session->cng_pt)) {
+                int sbytes = (int) bytes;
+                srtp_err_status_t stat = 0;
+
+                if (rtp_session->recv_ctx[rtp_session->srtp_idx_rtp]) {
+
+#ifdef DEBUG_SRTP_FLUSH
+                    if (sbytes) {
+                        q = 0;
+                        bytes_buf[0] = 0;
+                        for(i = 0; sbytes && i < sbytes; i++) {
+                            sprintf(bytes_buf + q, "0x%02x ", ((u_char *)(&rtp_session->recv_msg))[i]);
+                            q = strlen(bytes_buf);
+                        }
+                        switch_log_printf(SWITCH_CHANNEL_LOG, SWITCH_LOG_ERROR, "do_flush loop BEFORE decryption: bytes = %d\nData: %s\n", sbytes, bytes_buf);
+                    }
+#endif
+
+                    // Ensure we properly decrypt SRPT before reading DTMF
+                    if (!rtp_session->flags[SWITCH_RTP_FLAG_SECURE_RECV_MKI]) {
+                        stat = srtp_unprotect(rtp_session->recv_ctx[rtp_session->srtp_idx_rtp], &rtp_session->recv_msg.header, &sbytes);
+                    } else {
+                        stat = srtp_unprotect_mki(rtp_session->recv_ctx[rtp_session->srtp_idx_rtp], &rtp_session->recv_msg.header, &sbytes, 1);
+                    }
+
+#ifdef DEBUG_SRTP_FLUSH
+                    if (sbytes) {
+                        q = 0;
+                        bytes_buf[0] = 0;
+                        for(i = 0; sbytes && i < sbytes; i++) {
+                            sprintf(bytes_buf + q, "0x%02x ", ((u_char *)(&rtp_session->recv_msg))[i]);
+                            q = strlen(bytes_buf);
+                        }
+                        switch_log_printf(SWITCH_CHANNEL_LOG, SWITCH_LOG_ERROR, "do_flush loop AFTER decryption: bytes = %d\nData: %s\n", sbytes, bytes_buf);
+                    }
+#endif
+
+                    if (rtp_session->flags[SWITCH_RTP_FLAG_NACK] && stat == srtp_err_status_replay_fail) {
```

Comment:
is NACK being used with audio in your setup ?

Reply:
I'm not even sure.  I plagurised the code from else where, just to ensure do_flush would correctly decrypt DTMF.  Should I rip this out?